File: Dockerfile.go
# Multi-stage build for Go simulation server
# Final image < 30MB

FROM golang:1.21-alpine AS builder

RUN apk add --no-cache git

WORKDIR /app
COPY go/ ./go/
WORKDIR /app/go
RUN go mod download
RUN CGO_ENABLED=0 GOOS=linux go build -ldflags="-s -w" -o /server ./cmd/server

FROM alpine:3.19

RUN apk add --no-cache ca-certificates

COPY --from=builder /server /usr/local/bin/server

ENV GOGC=200

EXPOSE 9090

ENTRYPOINT ["server"]

File: Dockerfile.py
# Python trainer image with PyTorch + CUDA runtime

FROM pytorch/pytorch:2.2.0-cuda12.1-cudnn8-runtime

WORKDIR /app

RUN pip install --no-cache-dir grpcio protobuf gymnasium numpy requests redis boto3

COPY python/ ./python/
RUN pip install --no-cache-dir -e ./python/

COPY api/ ./api/

ENV PYTHONUNBUFFERED=1

ENTRYPOINT ["python", "-m"]
CMD ["rl_env_engine.learner"]
